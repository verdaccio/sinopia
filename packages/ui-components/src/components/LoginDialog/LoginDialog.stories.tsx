/* eslint-disable verdaccio/jsx-spread */
import { action } from '@storybook/addon-actions';
import React from 'react';

import LoginDialog from './LoginDialog';

export default {
  title: 'Components/Dialog/LoginDialog',
};

export const OpenLoginDialog = {
  render: (args) => <LoginDialog {...args} />,
  argTypes: {
    open: true,
  },
  args: {
    open: true,
    onClose: action('onClose'),
  },
};
