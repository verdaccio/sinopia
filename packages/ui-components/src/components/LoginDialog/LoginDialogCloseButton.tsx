import styled from '@emotion/styled';
import CloseIcon from '@mui/icons-material/Close';
import DialogTitle from '@mui/material/DialogTitle';
import IconButton from '@mui/material/IconButton';
import React from 'react';
import { useTranslation } from 'react-i18next';

import { Theme } from '../../';

const StyledIconButton = styled(IconButton)<{ theme?: Theme }>(({ theme }) => ({
  position: 'absolute',
  right: theme.spacing() / 2,
  top: theme.spacing() / 2,
  color: theme.palette.grey[500],
  zIndex: 99,
}));

interface Props {
  onClose: () => void;
}

const LoginDialogCloseButton: React.FC<Props> = ({ onClose }) => {
  const { t } = useTranslation();
  return (
    <DialogTitle>
      <StyledIconButton data-testid="close-login-dialog-button" onClick={onClose}>
        <CloseIcon titleAccess={t('button.close')} />
      </StyledIconButton>
    </DialogTitle>
  );
};

export default LoginDialogCloseButton;
