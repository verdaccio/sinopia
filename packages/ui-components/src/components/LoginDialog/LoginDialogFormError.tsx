import styled from '@emotion/styled';
import Error from '@mui/icons-material/Error';
import Box from '@mui/material/Box';
import SnackbarContent from '@mui/material/SnackbarContent';
import React, { memo } from 'react';

import { LoginError, Theme } from '../../';

const StyledSnackbarContent = styled(SnackbarContent)<{ theme?: Theme }>(({ theme }) => ({
  backgroundColor: theme.palette.error.dark,
  color: theme.palette.white,
}));

const StyledErrorIcon = styled(Error)<{ theme?: Theme }>(({ theme }) => ({
  fontSize: 20,
  opacity: 0.9,
  marginRight: theme.spacing(1),
}));

export interface FormValues {
  username: string;
  password: string;
}

interface Props {
  error: LoginError;
}

const LoginDialogFormError = memo(({ error }: Props) => {
  return (
    <StyledSnackbarContent
      message={
        <Box alignItems="center" data-testid="error" display="flex">
          <StyledErrorIcon />
          {error.description}
        </Box>
      }
    />
  );
});

export default LoginDialogFormError;
