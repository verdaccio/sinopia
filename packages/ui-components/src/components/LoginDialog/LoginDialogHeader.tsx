import styled from '@emotion/styled';
import CloseIcon from '@mui/icons-material/Close';
import LockOutlined from '@mui/icons-material/LockOutlined';
import Avatar from '@mui/material/Avatar';
import Box from '@mui/material/Box';
import IconButton from '@mui/material/IconButton';
import React from 'react';
import { useTranslation } from 'react-i18next';

import { Heading, Theme } from '../../';

interface Props {
  onClose?: () => void;
}

const LoginDialogHeader: React.FC<Props> = ({ onClose }) => {
  const { t } = useTranslation();

  return (
    <Box alignItems="center" display="flex" flexDirection="column" position="relative">
      {onClose && (
        <StyledIconButton aria-label={t('button.close')} onClick={onClose}>
          <CloseIcon />
        </StyledIconButton>
      )}
      <StyledAvatar>
        <LockOutlined />
      </StyledAvatar>
      <Heading>{t('button.login')}</Heading>
    </Box>
  );
};

export default LoginDialogHeader;

const StyledAvatar = styled(Avatar)<{ theme?: Theme }>(({ theme }) => ({
  margin: theme.spacing(1),
  backgroundColor:
    theme.palette.mode === 'light' ? theme.palette.primary.main : theme.palette.cyanBlue,
  color: theme.palette.white,
}));

const StyledIconButton = styled(IconButton)<{ theme?: Theme }>(({ theme }) => ({
  position: 'absolute',
  right: theme.spacing() / 2,
  top: theme.spacing() / 2,
  color: theme.palette.grey[500],
}));
