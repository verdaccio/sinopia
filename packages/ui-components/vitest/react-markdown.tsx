import React from 'react';
import { vi } from 'vitest';

// Mock react-markdown globally
vi.mock('react-markdown', () => ({
  default: ({ children }: { children: React.ReactNode }) => <>{children}</>,
}));
