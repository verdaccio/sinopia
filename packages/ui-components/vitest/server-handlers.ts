import fs from 'fs';
import { HttpResponse, http } from 'msw';
import path from 'path';

export const handlers = [
  http.get('http://localhost:9000/-/verdaccio/data/packages', () => {
    return HttpResponse.json(Array(400).fill(require('./api/home-packages.json')));
  }),

  http.get('http://localhost:9000/-/verdaccio/data/sidebar/storybook', ({ params }) => {
    const { v } = params;
    if (v) {
      return HttpResponse.json(require('./api/storybook-v.json'));
    } else {
      return HttpResponse.json(require('./api/storybook-sidebar.json'));
    }
  }),

  http.get('http://localhost:9000/-/verdaccio/data/search/*', () => {
    return HttpResponse.json(require('./api/search-verdaccio.json'));
  }),

  http.get('http://localhost:9000/-/verdaccio/data/package/readme/storybook', () => {
    return HttpResponse.text(require('./api/storybook-readme')());
  }),

  http.get('http://localhost:9000/-/verdaccio/data/sidebar/jquery', () => {
    return HttpResponse.json(require('./api/jquery-sidebar.json'));
  }),
  http.get('http://localhost:9000/-/verdaccio/data/sidebar/JSONStream', () => {
    return new HttpResponse('unauthorized', { status: 401 });
  }),
  http.get('http://localhost:9000/-/verdaccio/data/sidebar/semver', () => {
    return new HttpResponse('internal server error', { status: 500 });
  }),
  http.get('http://localhost:9000/-/verdaccio/data/sidebar/kleur', () => {
    return new HttpResponse('not found', { status: 404 });
  }),
  http.get('http://localhost:9000/-/verdaccio/data/sidebar/glob', () => {
    return HttpResponse.json(require('./api/glob-sidebar.json'));
  }),
  http.get('http://localhost:9000/-/verdaccio/data/package/readme/glob', () => {
    return HttpResponse.text('foo glob');
  }),
  http.get('http://localhost:9000/-/verdaccio/data/sidebar/got', () => {
    return HttpResponse.json(require('./api/got-sidebar.json'));
  }),
  http.get('http://localhost:9000/-/verdaccio/data/package/readme/got', () => {
    return HttpResponse.text('foo got');
  }),
  http.get('http://localhost:9000/-/verdaccio/data/package/readme/jquery', () => {
    return HttpResponse.text(require('./api/jquery-readme')());
  }),
  http.get('http://localhost:9000/verdaccio/-/verdaccio-1.0.0.tgz', () => {
    const fileName = path.resolve(__dirname, './api/verdaccio-1.0.0.tgz');
    const fileContent = fs.readFileSync(fileName);
    return new HttpResponse(fileContent, {
      status: 200,
      headers: { 'Content-Type': 'application/octet-stream' },
    });
  }),

  http.post<{ username: string; password: string }, { token: string; username: string }>(
    'http://localhost:9000/-/verdaccio/sec/login',
    async ({ request }) => {
      const body = await request.json();
      if (body.username === 'fail') {
        return new HttpResponse('unauthorized', { status: 401 });
      }

      return HttpResponse.json({ username: body.username, token: 'valid token' });
    }
  ),
];
