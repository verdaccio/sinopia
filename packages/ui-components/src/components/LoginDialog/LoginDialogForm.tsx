/* eslint-disable verdaccio/jsx-spread */
import styled from '@emotion/styled';
import Button from '@mui/material/Button';
import React, { memo } from 'react';
import { useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';

import { LoginError, TextField, Theme } from '../../';
import LoginDialogFormError from './LoginDialogFormError';

const StyledForm = styled('form')<{ theme?: Theme }>(({ theme }) => ({
  marginTop: theme.spacing(1),
}));

const StyledButton = styled(Button)<{ theme?: Theme }>(({ theme }) => ({
  margin: theme.spacing(3, 0, 2),
}));

export interface FormValues {
  username: string;
  password: string;
}

interface Props {
  onSubmit: (formValues: FormValues) => void;
  error?: LoginError;
}

const LoginDialogForm = memo(({ onSubmit, error }: Props) => {
  const { t } = useTranslation();
  const {
    register,
    handleSubmit,
    formState: { isValid, errors },
  } = useForm<FormValues>({ mode: 'onChange' });

  const onSubmitForm = (formValues: FormValues) => {
    onSubmit(formValues);
  };

  return (
    <StyledForm noValidate={true} onSubmit={handleSubmit(onSubmitForm)}>
      <TextField
        autoComplete="username"
        error={!!errors.username}
        fullWidth={true}
        helperText={errors.username?.message}
        id="login--dialog-username"
        {...register('username', {
          required: { value: true, message: t('form-validation.required-field') },
        })}
        label={t('form.username')}
        margin="normal"
        name="username"
        placeholder={t('form-placeholder.username')}
        required={true}
        variant="outlined"
      />
      <TextField
        autoComplete="current-password"
        error={!!errors.password}
        fullWidth={true}
        helperText={errors.password?.message}
        id="login--dialog-password"
        {...register('password', {
          required: { value: true, message: t('form-validation.required-field') },
        })}
        data-testid="password"
        label={t('form.password')}
        margin="normal"
        name="password"
        placeholder={t('form-placeholder.password')}
        required={true}
        type="password"
        variant="outlined"
      />
      {error && <LoginDialogFormError error={error} />}
      <StyledButton
        color="primary"
        data-testid="login-dialog-form-login-button"
        disabled={!isValid}
        fullWidth={true}
        id="login--dialog-button-submit"
        size="large"
        type="submit"
        variant="contained"
      >
        {t('button.login')}
      </StyledButton>
    </StyledForm>
  );
});

export default LoginDialogForm;
