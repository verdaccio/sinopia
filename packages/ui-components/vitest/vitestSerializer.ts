import { SnapshotSerializer } from 'vitest';

const serializer: SnapshotSerializer = {
  serialize(val, config, indentation, depth, refs, printer) {
    for (const className of [...val.classList]) {
      if (className.startsWith('css-')) {
        const hashEnd = className.indexOf('-', 4);
        if (hashEnd >= 0) {
          val.classList.replace(className, 'makeStyles' + className.substring(hashEnd));
        } else {
          val.classList.remove(className);
        }
      }
    }
    if (!val.classList.length) {
      val.removeAttribute('class');
    }
    return printer(val, config, indentation, depth, refs);
  },
  test(val) {
    return val instanceof Element && [...val.classList].some((name) => name.startsWith('css-'));
  },
};

/**
 * An object snapshot serializer that removes dynamic CSS classes (generated by makeStyle) from DOM elements.
 *
 * `css-hash-suffix` is transformed into `makeStyles-suffix`
 * `css-hash` is removed
 */
export default serializer;
