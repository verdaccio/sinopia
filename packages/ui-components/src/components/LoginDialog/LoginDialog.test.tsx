import React from 'react';
import { vi } from 'vitest';

import { api, store } from '../../';
import {
  act,
  cleanup,
  fireEvent,
  renderWithStore,
  screen,
  waitFor,
} from '../../test/test-react-testing-library';
import LoginDialog from './LoginDialog';

describe('<LoginDialog /> component', () => {
  beforeEach(() => {
    vi.resetModules();
    vi.resetAllMocks();
    cleanup();
  });

  test('should render the component in default state', () => {
    const props = {
      onClose: vi.fn(),
    };
    const { container } = renderWithStore(<LoginDialog onClose={props.onClose} />, store);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should load the component with the open prop', async () => {
    const props = {
      open: true,
      onClose: vi.fn(),
    };

    const { getByTestId } = renderWithStore(
      <LoginDialog onClose={props.onClose} open={props.open} />,
      store
    );

    const loginDialogHeading = await waitFor(() => getByTestId('login-dialog-form-login-button'));
    expect(loginDialogHeading).toBeTruthy();
  });

  test('onClose: should close the login modal', async () => {
    const props = {
      open: true,
      onClose: vi.fn(),
    };

    const { getByTestId } = renderWithStore(
      <LoginDialog onClose={props.onClose} open={props.open} />,
      store
    );

    const loginDialogButton = await waitFor(() => getByTestId('close-login-dialog-button'));
    expect(loginDialogButton).toBeTruthy();

    await act(() => {
      fireEvent.click(loginDialogButton, { open: false });
    });

    expect(props.onClose).toHaveBeenCalled();
  });

  test('setCredentials - should set username and password in state', async () => {
    const props = {
      open: true,
      onClose: vi.fn(),
    };

    vi.spyOn(api, 'request').mockImplementation(() =>
      Promise.resolve({
        username: 'xyz',
        token: 'djsadaskljd',
      })
    );

    await act(async () => {
      renderWithStore(<LoginDialog onClose={props.onClose} open={props.open} />, store);
    });

    const userNameInput = screen.getByPlaceholderText('form-placeholder.username');
    expect(userNameInput).toBeInTheDocument();

    fireEvent.focus(userNameInput);

    fireEvent.change(userNameInput, { target: { value: 'xyz' } });

    const passwordInput = screen.getByPlaceholderText('form-placeholder.password');
    expect(passwordInput).toBeInTheDocument();
    fireEvent.focus(passwordInput);

    await act(async () => {
      fireEvent.change(passwordInput, { target: { value: '1234' } });
    });
    const signInButton = screen.getByTestId('login-dialog-form-login-button');
    expect(signInButton).not.toBeDisabled();

    await act(async () => {
      fireEvent.click(signInButton);
    });
    expect(props.onClose).toHaveBeenCalledTimes(1);
  });

  test.todo('validateCredentials: should validate credentials');

  test.todo('submitCredentials: should submit credentials');
});
