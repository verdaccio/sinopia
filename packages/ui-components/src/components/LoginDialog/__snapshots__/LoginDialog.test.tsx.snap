// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<LoginDialog /> component > should render the component in default state 1`] = `null`;
